### DISCUSSION TITLE
RP2040 - Performance is NOT mediocre (spoiler: it's very fast)

### DISCUSSION THREAD

**Discussion by SezaroSteve** [ZuluSCSI/ZuluSCSI-firmware#133] [General]
Hello everybody,

There are opinions floating around on the interwebs that the RP2040's "Harddisk" performance is only "so so". Most of the time, these opinions are formed because of poor SCSI Implementations on the host system or because people still have numbers of older SCSI2SD solutions in their head.
So in their heads, ZuluSCSI is not much of an improvement. I beg to differ and I can back that up with hard data (see what I did there ;-)

I use the "normal/regular" RP2040 (not the "mini") on an Amiga 1200, accelerated with a Phase5 Blizzard 1240T (CPU = 68040 @ 40Mhz) with the accompanying "SCSI-Kit Mk IV", which is a very well implemented SCSI-2 Implementation with DMA (Direct Memory Access).
RP2040 Firmware 1.2.0 solves the write-performance issue and the Zulu now flies. It is stonki'n fast.

The SD-Card I prefer to use is the "SanDisk Extreme Pro 64GB 170MB/s" which is a full-size SDXC card. It delivers better, and especially (and more important) consistent performance as many other cards I used. It does not "bog down" under heavy writing.

Here are the numbers
```
Amiga 1200
2MB Chip
Phase5 Blizzard 1240T @ 40Mhz, 64MB FastMem
Kickstart 3.1
Workbench 3.1
1230scsi.device Version 8.5

MKSoft ScsiSpeed 4.2  Copyright © 1989-92 MKSoft Development
------------------------------------------------------------
CPU: 68040  AmigaOS Version: 40.68  Normal Video DMA
Device: 1230scsi.device:1
Comments: ZuluSCSI RP240 FW 1.2.0 , SCSI Mode = SCSI-2 Sync

CPU Speed Rating: 5693

Testing with a 512 byte, MEMF_FAST, LONG-aligned buffer.
Read from SCSI:    501045 bytes/sec  |  CPU Available: 41%

Testing with a 4096 byte, MEMF_FAST, LONG-aligned buffer.
Read from SCSI:   1876787 bytes/sec  |  CPU Available: 72%

Testing with a 32768 byte, MEMF_FAST, LONG-aligned buffer.
Read from SCSI:   5210112 bytes/sec  |  CPU Available: 91%

Testing with a 262144 byte, MEMF_FAST, LONG-aligned buffer.
Read from SCSI:   5837003 bytes/sec  |  CPU Available: 98%
```

SysInfo shows 5.57 MB/s (it does Raw Reads).
This is with Sync SCSI-2. But even under Async SCSI-2, I still get 4.41 MB/s.

Write performance is not measured by this ScsiSpeed tool. But SysSpeed for example, has a write benchmark. Result: 3.43 MB/s which is really good and, most important, it is consistent. It will do it all day long without slowing to a crawl. Using the right SD-Card is very important for this. Do NOT cheap out on the SD-Card. It plays an important role.

What is important is the "CPU Available" number. As you can see, it is very high, meaning the CPU Load caused by the SCSI I/O's is very low. This frees the CPU up for other tasks. This makes the machine feel very snappy and responsive.
You can get similar high throughput with an IDE CF-Card / SD-Card implementation on say, a Terrible Fire 536, but the CPU load is very high, almost choking the CPU. Old school IDE is nowhere near as efficient. It bogs down the CPU much more und high I/O load. People are often focused on what SysInfo says (on the Amiga in this case). But that's only a raw-read with a rather large I/O Blocksize. In real life, this is largely irrelevant. Being able to handle little I/O's fast and efficiently is more important in real-life applications.

In my opinion, the SCSI virtualization implementation in the ZuluSCSI RP2040 solution is starting to reach a maturity level, well worth anybodies attention if you are looking for, or needing to, replacing a real SCSI-1 or SCSI-2 Hard-drive with something less "spinny".

Update:  In an Amiga 3000D with a WarpEngine 3040 accelerator and using it's onboard SCSI (so not the A3000 native SCSI), the ZuluSCSI churns out a consistent 9.34 MB/s !!

The same inside an A4000T with a BFG9060 @ 50Mhz does 8.53 MB/s

So if you are experiencing slow performance, trust me, the bottleneck it is NOT the ZuluSCSI (assuming you use a decent SD-Card and not some cheap AliExpress knock-off).

**Comment by makoovka** [comment]
Thank you for sharing it ! Did you have a chance to compare ZuluSCSI RP2040 full sized version with it's mini counterpart? I wonder if the performance is the same. I own Blizzard 1230 with the same SCSI controller and was thinking of putting ZuluSCSI RP2040 Mini inside A1200's case. Or do you think that regular version would fit inside too? If yes I could spare few bucks :)

**Reply by PetteriAimonen** [reply]
ZuluSCSI RP2040 at any size should have similar performance, the SCSI hardware and CPU is identical. The mini versions support only MicroSD cards, but it seems they are as fast as full size SD cards nowadays.

**Comment by SezaroSteve** [comment]
In all my ZuluSCSI's (I have 4), the SD cards I use are actually MicroSD cards in a classic SD-Card Adapter. So removing that adapter and putting the MicroSD card into the MicroSD card of a RP2040 Mini, should yield the same results if all the other hardware, incl. the card, is the same and only the SD-Card-slot size is different (mini vs. regular).

**Comment by Konf** [comment]
Hi! 
What kind of write speeds do you have with A4000T onboard controller?
I have perfect ~9 MB/s reads and ~4.5 MB/s writes with my setup in SysSpeed (PFS3 partition).
Wonder if something can be done to speed up writes a little bit (maybe some tweaking?), or that's it's current limit.

**Reply by aperezbios** [reply]
@Konf We've already made significant write performance improvements via optimization with the ZuluSCSI RP2040, since it was released in November of 2022. We'd be happy to accept a pull request to improve write performance, though :) Anybody can work on this if they so desire, but it won't be easy to improve things on that front.

**Reply by SezaroSteve** [reply]
Ok, I ran a couple of benchmarks just now. Cold boot into Workbench, started SysSpeed and its Drive benchmark and the first reading was taken. So no caching was used in any form (things obviously get quicker when you give buffers/caches the chance to warm up, but that is not a good measurement of a device's raw, uncached performance).

Environment:
ROM 3.1.4 in all machines.
FileSystem: PFS3 normal mode (no SCSI Direct mode as it destroys my data).
ZuluSCSI RP2040 with current firmware.
SD-Cards are all high-grade Sandisk Extreme Pro 64GB cards formatted as EXFAT with 4 image-files as virtual harddisks (DH0 to DH3).
SCSI ID 6 is used as the virtual CD-ROM drive.
SCSI ID 5 is not used by the Zulu as that's the ID of the external Iomega ZIP-Drive when attached (I don't use it often and it was not hooked up during these benchmarks).
SysSpeed 2.6 Benchmarks, values for "Drive ReadFile" and "Drive WriteFile".

Results:
My A1200 with Blizzard 1240T @ 40MHz and the Blizzard's own SCSI-controller:  5.61 MB/s read and 4.34 MB/s write.
My A4000T with BFG9060 @ 50MHz and the onboard SCSI-controller:  8.45 MB/s read and 6.21 MB/s write.
My A3000D with WarpEngine 3040 @ 40MHz and the WarpDrive SCSI-controller:  9.38 MB/s read and 6.52 MB/s write.

(this cleary shows how good the architecture of the WarpEngine and it's SCSI implementation is).
